formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoding.kt

```diff
@@ -46,165 +48,367 @@ private const val SINGLE_PRECISION_MAX_EXPONENT = 0xFF
+                descriptor = descriptor,
+                data = ByteArrayOutput().apply { write(BREAK) },
+                next = null,
+            )
+            currentToken = currentToken.next!!
+        }
+
     }
 
-    override fun endStructure(descriptor: SerialDescriptor) = encoder.end()
 
     override fun encodeElement(descriptor: SerialDescriptor, index: Int): Boolean {
+        val parent = structureStack.peek() as Token?
+        val name = descriptor.getElementName(index) as String?
+        val label = descriptor.getCborLabel(index)
+
+        val preamble = encodePreamble(parent?.descriptor, index, label, name)
+
         encodeByteArrayAsByteString = descriptor.isByteString(index)
-        val name = descriptor.getElementName(index)
-        encoder.encodeString(name)
+        currentToken.next = Token(
+            descriptor = descriptor.getElementDescriptor(index),
+            data = null,
+            next = null,
+            preamble = preamble
+        )
+        currentToken = currentToken.next!!
+        runCatching { structureStack.peek().numChildren = structureStack.peek().numChildren!! + 1 }.exceptionOrNull()
+            ?.printStackTrace()
         return true
     }
 
-    override fun encodeString(value: String) = encoder.encodeString(value)
+    private fun encodePreamble(
+        parentDescriptor: SerialDescriptor?,
+        index: Int,
+        label: Long?,
+        name: String?
+    ): ByteArrayOutput {
+        val output = ByteArrayOutput()
+
+        parentDescriptor?.let { descriptor ->
+            if (!descriptor.hasArrayTag()) {
+                if (cbor.writeKeyTags) {
+                    index.let { descriptor.getKeyTags(it)?.forEach { output.encodeTag(it) } }
+                }
+                if ((descriptor.kind !is StructureKind.LIST) && (descriptor.kind !is StructureKind.MAP) && (descriptor.kind !is PolymorphicKind)) {
+                    //indices are put into the name field. we don't want to write those, as it would result in double writes
+                    if (cbor.preferCborLabelsOverNames && label != null) {
+                        output.encodeNumber(label)
+                    } else if (name != null) {
+                        output.encodeString(name)
+                    }
+                }
+            }
+        }
+        if (cbor.writeValueTags) {
+            index.let { parentDescriptor?.getValueTags(it)?.forEach { output.encodeTag(it) } }
+        }
+        return output
+    }
 
-    override fun encodeFloat(value: Float) = encoder.encodeFloat(value)
-    override fun encodeDouble(value: Double) = encoder.encodeDouble(value)
+    //If any of the following functions are called for serializing raw primitives (i.e. something other than a class,
+    // list, map or array, no children exist and the root node needs the data
+    override fun encodeString(value: String) {
+        currentToken.data = ByteArrayOutput().also { it.encodeString(value) }
```

Comment:
Creating `ByteArrayOutput()` only to encode a single integer or a string is definitely going to slow down encoding, because you are creating a lot of small byte arrays instead of a single large one. Also, it will result in copying each byte twice (once encoding here, second time copying from token to global output).

I've wrote a simple benchmark and compared your branch against `dev`:

Dev:
```
Benchmark                Mode  Cnt     Score     Error   Units
CborBaseline.fromBytes  thrpt   10   727.284 ± 106.661  ops/ms
CborBaseline.toBytes    thrpt   10  1896.529 ±  40.447  ops/ms
```

Your branch:
```
Benchmark                Mode  Cnt     Score    Error   Units
CborBaseline.fromBytes  thrpt   10   661.809 ± 65.200  ops/ms
CborBaseline.toBytes    thrpt   10  1208.825 ± 45.491  ops/ms
```
(benchmark patch: https://gist.github.com/sandwwraith/2ebed6c997264c730d880ca4ef75bd46)

As you can see, while decoding performance is roughly the same (or at least close), encoding performance has dropped significantly. I did not profile the code, but I expect one of the main reasons is fragmented data in arrays. (as benchmark doesn't use any new features).

Reply:
If I understood code correctly, all this hassle with data manipulation is required for correct computing collection sizes in `endStructure` (because you want to write definite lengths, you need first to process all elements and then start writing the array). 

If that is the only reason, I suggest looking into `override fun beginCollection(descriptor: SerialDescriptor, collectionSize: Int): CompositeEncoder` from `Encoder` interface, which passes you the collection size in advance. It looks like a perfect fit here, and you'll be able to put everything into a single output.

There's a caveat, though: serializer implementation should explicitly call it instead of `beginStructure`. All standard collection serializers do that, so you're safe here. If, on some rare occasion, you'll encounter a custom serializer that has `LIST` kind, but does not call `beginCollection`, you have two options:

1. Say that you always write such things as indefinite length (even with `writeDefiniteLengths = true`). IMO, this behavior is reasonable, as it is a custom serializer's author oversight.
2. Only in this rare case, put everything into an intermediate `ByteArrayOutput` to append it later to a general one. It is absolutely crucial that the normal process of encoding won't be affected by this. Implementing this will require some careful handling here and there, but seems manageable if you absolutely need definite lengths everywhere.

Reply:
Thanks! That sounds perfect, but it will require the documentation to go along with the rewrite.

The thing is: COSE absolutely requires definite length. That's just how the spec is written and there's no wiggle room. None of what we pushed here was done because we think it to be a nice addition but because [RFC 8812](https://datatracker.ietf.org/doc/html/rfc8812) demands it.<br> Still, I think `override fun beginCollection(descriptor: SerialDescriptor, collectionSize: Int): CompositeEncoder` from `Encoder` is perfect and we'll leave the sloppy case you described as a (well-documented) limitation for later (unless we notice that it bites us, already ;-))

Reply:
It seems, I was overly optimistic. It works for collections, but not for other structures (instances of arbitrary classes, to that won't work). I guess it's back to a two-pass encoder…

Reply:
I think for regular `StructureKind.CLASS` classes, you just want to use `SerialDescriptor.elementsCount`

Reply:
we'll do deferred encoding. next iteration of the PR will be ready next week.

Reply:
if you want to peek, we've already made it way leaner (but still two-pass encoding, as we need to be able to enforce definite length encoding in all cases for CBOR): https://github.com/a-sit-plus/kotlinx.serialization/tree/tmp/updateCoseUpstream

Reply:
Are there any other problems? `beginCollection` covers all collections, and `SerialDescriptor.elementsCount` covers the rest.

Reply:
I seem to recall that I tried that during early experiments and that nullable properties can throw off `SerialDescriptor.elementsCount`. But that was months ago, so I'll try.

Reply:
@sandwwraith I really appreciate the help, thank you very much! But sadly, no dice!

When setting `endcodeDefaults` to false, everything falls apart. See the [test case here](https://github.com/a-sit-plus/kotlinx.serialization/blob/tmp/proofOfFail/formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborDefLenTest.kt), which demonstrates this problem. I've just added `println()` statements to the encoder:

 * on two occasions using the functions you mentioned (which produce the wrong number of elements in a structure, if `encodeDefaults` is set to false
 * once using our two-pass approach (which now directly uses the passed `ByteArrayOutput` and no more copying, but writing encoding).

Sample of problematic output of the test case looks as follows (annotated here, to highlight the issue):

```kotlin
ENCODE DEFAULTS=FALSE, ALL SET

// **wrong. the value set corresponds to the default and shall be omitted**
BUILT-IN START CLASS		of size 9 for kotlinx.serialization.cbor.TypesUmbrella
// **wrong. the value set corresponds to the default and shall be omitted**

BUILT-IN START COLLECTION	of size 2 for kotlin.collections.ArrayList
BUILT-IN START COLLECTION	of size 2 for kotlin.collections.LinkedHashMap //would need to check if it is a map, then multiply by two. Ignore it here, that's not the point
BUILT-IN START CLASS		of size 1 for kotlinx.serialization.cbor.Simple
BUILT-IN START COLLECTION	of size 1 for kotlin.collections.ArrayList
BUILT-IN START CLASS		of size 1 for kotlinx.serialization.cbor.Simple
BUILT-IN START COLLECTION	of size 2 for kotlin.ByteArray


// **correctly calculated, since defaults are ommitted**
A-SIT PLUS PATCH calculated STRUCT/COLL: of size 8 for kotlinx.serialization.cbor.TypesUmbrella
// **correctly calculated, since defaults are ommitted**

A-SIT PLUS PATCH calculated STRUCT/COLL: of size 2 for kotlin.collections.ArrayList
A-SIT PLUS PATCH calculated STRUCT/COLL: of size 4 for kotlin.collections.LinkedHashMap
A-SIT PLUS PATCH calculated STRUCT/COLL: of size 1 for kotlinx.serialization.cbor.Simple
A-SIT PLUS PATCH calculated STRUCT/COLL: of size 1 for kotlin.collections.ArrayList
A-SIT PLUS PATCH calculated STRUCT/COLL: of size 1 for kotlinx.serialization.cbor.Simple
A-SIT PLUS PATCH calculated STRUCT/COLL: of size 2 for kotlin.ByteArray
```

I am also confident to find more issues. So at the moment if feels like that, yes, it can be made to work, but the code would be really messy in the end. (Also, yes, [we absolutely need the `endocdeDefaults=false`](https://github.com/a-sit-plus/kmp-crypto/blob/00034fff0a3909ecce1978f698273c9387f452e5/datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKey.kt#L30) feature for COSE, otherwise there is no way to be RFC-compliant. I can elaborate more on COSE, too, if required).

So we'll be polishing the two-pass deferred encoding approach further and we'll rebase everything to our PR branch once ready. The two-pass encoding is already much more comprehensible and streamlined, thanks to your feedback.